cdap-ui/app/cdap/components/PluginSchemaEditor/index.tsx

```diff
@@ -0,0 +1,539 @@
+    schemas: this.props.schemas,
+    loading: false,
+    mode: this.getMode(),
+    schemaRowCount: null,
+  };
+
+  private ee = ee(ee);
+
+  constructor(props) {
+    super(props);
+    if (!this.props.disabled) {
+      this.ee.on('schema.import', this.onSchemaImport);
+      this.ee.on('dataset.selected', this.onSchemaImport);
+    }
+    const doesActionDropdownHasExport = Object.keys(this.props.actionsDropdownMap).find(
+      (key) => key.toLowerCase() === 'export'
+    );
+    if (doesActionDropdownHasExport) {
+      // Schema can be exported on detailed view even if it is disabled.
+      this.ee.on('schema.export', this.onSchemaExport);
+    }
+    window.addEventListener('resize', this.calculateSchemaRowCount);
+  }
+
+  public componentWillReceiveProps(nextProps: IPluginSchemaEditorProps) {
+    if (!Object.keys(nextProps.actionsDropdownMap).length) {
+      this.actions = [];
+      return;
+    }
+    this.actions = Object.values(nextProps.actionsDropdownMap).map((value) => value);
+    return;
+  }
+
+  public shouldComponentUpdate(nextProps: IPluginSchemaEditorProps, nextState) {
+    const { disabled, isSchemaMacro, actionsDropdownMap, schemas } = nextProps;
+    const { schemaRowCount, loading, mode, error } = nextState;
+    const newActions = Object.values(actionsDropdownMap)
+      .filter((value) => typeof value === 'string')
+      .join('--');
+    const existingActions = Object.values(this.props.actionsDropdownMap)
+      .filter((value) => typeof value === 'string')
+      .join('--');
+    const existingSchemas = this.props.schemas.map((s) => s.schema).join('__');
+    const newSchemas = schemas.map((s) => s.schema).join('__');
+
+    const didPropsChange =
+      disabled !== this.props.disabled ||
+      isSchemaMacro !== this.props.isSchemaMacro ||
+      newActions !== existingActions ||
+      newSchemas !== existingSchemas;
+
+    const didStateChange =
+      schemaRowCount !== this.state.schemaRowCount ||
+      loading !== this.state.loading ||
+      mode !== this.state.mode ||
+      error !== this.state.error;
+    return didStateChange || didPropsChange;
+  }
+
+  public componentDidMount() {
+    this.calculateSchemaRowCount();
+  }
+
+  public componentWillUnmount() {
+    this.ee.off('schema.import', this.onSchemaImport);
+    this.ee.off('schema.export', this.onSchemaExport);
+    this.ee.on('dataset.selected', this.onSchemaImport);
+    window.removeEventListener('resize', this.calculateSchemaRowCount);
+  }
+
+  public calculateSchemaRowCount = () => {
+    if (this.containerRef) {
+      const { height } = this.containerRef.getBoundingClientRect();
+      let schemaRowCount = Math.floor(height / FieldsListBase.heightOfRow) - 1;
+      // For sinks we can't determine the height. So set it to by default 20 (20 * 34 = 680px in height)
+      if (schemaRowCount < 5) {
+        schemaRowCount = 20;
+      }
+      this.setState({ schemaRowCount });
+    }
+  };
+
+  public onSchemaExport = () => {
```

Comment:
We have the DownFile utility function you could use here (possibly with a small tweak for the export file name).

Reply:
This is a valid point. Our DownloadFile API is not right. It has hard requirement of naming of the file being in the exported config which is not the case all the time. I will do this outside of this PR to change the helper function and change all uses of it. Don't want to introduce a huge change in this PR at this time.

Reply:
That's fair. Can we add a TO DO comment here as a reminder?